tracer/src/Datadog.Trace/ClrProfiler/CallTarget/Handlers/IntegrationMapper.cs

```diff
@@ -123,24 +123,44 @@ internal static DynamicMethod CreateBeginMethodDelegate(Type integrationType, Ty
                     targetParameterTypeConstraint = targetParameterType.GetGenericParameterConstraints().FirstOrDefault(pType => pType != typeof(IDuckType));
                     if (targetParameterTypeConstraint is null)
                     {
-                        callGenericTypes.Add(sourceParameterType);
+                        callGenericTypes.Add(targetParameterType.IsByRef ? sourceParameterType : sourceParameterType.GetElementType());
                     }
                     else
                     {
-                        var result = DuckType.GetOrCreateProxyType(targetParameterTypeConstraint, sourceParameterType);
+                        var result = DuckType.GetOrCreateProxyType(targetParameterTypeConstraint, sourceParameterType.GetElementType());
                         parameterProxyType = result.ProxyType;
                         callGenericTypes.Add(parameterProxyType);
                     }
                 }
-                else if (!targetParameterType.IsAssignableFrom(sourceParameterType) && (!(sourceParameterType.IsEnum && targetParameterType.IsEnum)))
+                else
                 {
-                    throw new InvalidCastException($"The target parameter {targetParameterType} can't be assigned from {sourceParameterType}");
+                    var srcParameterType = sourceParameterType.IsByRef ? sourceParameterType.GetElementType() : sourceParameterType;
+                    var trgParameterType = targetParameterType.IsByRef ? targetParameterType.GetElementType() : targetParameterType;
+
+                    if (!trgParameterType.IsAssignableFrom(srcParameterType) && (!(srcParameterType.IsEnum && trgParameterType.IsEnum)))
+                    {
+                        throw new InvalidCastException($"The target parameter {targetParameterType} can't be assigned from {sourceParameterType}");
+                    }
```

Comment:
I'd love to create some analyzers to help us with this sort of thing. If I understand correctly:

* You **must** use `ref`/`out` for an integration if the original method uses `ref`/`out`
* You **can** use `ref` for integrations that don't use `ref`, as long as you don't use automatic generic-constraint duck typing
  * Should we _prefer_ using `ref` in general? Or keep it only in special cases, for large structs?

Unrelated to this change, but additional analyzers we could do:
* Prefer virtual class duck types over interface where possible
* Prefer struct duck types over class/interface where possible

Reply:
All correct!

> Should we prefer using ref in general? Or keep it only in special cases, for large structs?

Internally all is passed by ref now, so yes we could change all integrations to use `ref`. I kind of prefer to keep it optional, when an integration doesn't have the `ref` keyword we are emiting a `LDOBJ` to load the value of the address before calling the integration. In the general case I don't have any problem on doing that. As you said, this is different for large structs where I have the strong preference to force the `ref` on the integration.